packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,241 @@
+import { StaticDataView, sizeOfUTF8 } from './data-view';
+import { fastStartsWith } from './utils';
+
+export type EnvKeys =
+  | 'ext_ghostery'
+  | 'ext_abp'
+  | 'ext_ublock'
+  | 'ext_ubol'
+  | 'ext_devbuild'
+  | 'env_chromium'
+  | 'env_edge'
+  | 'env_firefox'
+  | 'env_mobile'
+  | 'env_safari'
+  | 'env_mv3'
+  | 'false'
+  | 'cap_html_filtering'
+  | 'cap_user_stylesheet'
+  | 'adguard'
+  | 'adguard_app_android'
+  | 'adguard_app_ios'
+  | 'adguard_app_mac'
+  | 'adguard_app_windows'
+  | 'adguard_ext_android_cb'
+  | 'adguard_ext_chromium'
+  | 'adguard_ext_edge'
+  | 'adguard_ext_firefox'
+  | 'adguard_ext_opera'
+  | 'adguard_ext_safari'
+  | (string & {});
+
+export class Env extends Map<EnvKeys, boolean> {}
+
+export const enum PreprocessorTypes {
+  INVALID = 0,
+  BEGIF = 1,
+  ELSE = 2,
+  ENDIF = 3,
+}
+
+export function detectPreprocessor(line: string) {
+  if (line.charCodeAt(1) !== 35 /* '#' */) {
+    return PreprocessorTypes.INVALID;
+  }
+
+  const command = line.slice(2);
+
+  if (fastStartsWith(command, 'if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (command === 'else') {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (command === 'endif') {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const tokenizerPattern = /(!|&&|\|\||\(|\)|[a-zA-Z0-9_]+)/g;
+
+const tokenize = (expression: string) => expression.match(tokenizerPattern);
+
+const precedence: Record<string, number> = {
+  '!': 2,
+  '&&': 1,
+  '||': 0,
+};
```

Comment:
Nit. To make the type even stricter.
```suggestion
const precedence = {
  '!': 2,
  '&&': 1,
  '||': 0,
} as const;
```

Reply:
It was to make the object less-stricter to avoid a type issue in the following reference: https://github.com/ghostery/adblocker/pull/3782/files#diff-9266c61212d0a9d9793cdd6255b30de0e6dcf7645494ee76f87e1c39f23a1682R129